### ISSUE TITLE
[feature] Pretty print perl data structures using a single command

### ISSUE THREAD

**Issue by rai-gaurav** [Perl/perl5#18455]
Right now if we want to see the internals of any Perl data structure, we use Dumper;

```
use Data::Dumper;

print Dumper(\@array);
print Dumper(\%hash);
print Dumper($scalar);
```
While this is good and get the job done, if we want to do something quickly and see the result with just print we will get array or hash ref. After that we will realize we have to include the dumper and pass the array ref or hash ref. 

It would be good we can dump the data structure with a single command.

**Describe the solution you'd like**

I like the pprint keyword in Python but there also you have to import it before using it.
For Perl if there is a keyword something like that can pretty print my data structure without passing reference it would be good especially during debug time. E.g.

```
pprint @array;
[  
    'abc',
    'ijk',
    'xyz'
]

pprint %hash  or   pprint  $output_fh  %hash;
{
    'key1' => 'value1',
    'key2' => 'value2'
}

pprint $scalar;
abc

```

I have just use the 'pprint' keyword as an example. It can be anything(e.g. 'printd' ?).

**Comment by rai-gaurav** [comment]
I am not sure about whether this came in past or not. Neither I have any idea about the complexity of the implementation.
Just a thought.

**Comment by jkeenan** [comment]
On 1/4/21 9:26 AM, sudo-batman wrote:
> Right now if we want to see the internals of any Perl data structure, we 
> use Dumper;
> 
> |use Data::Dumper; print Dumper(\@array); print Dumper(\%hash); print 
> Dumper($scalar); |
> 
> While this is good and get the job done, if we want to do something 
> quickly and see the result with just print we will get array or hash 
> ref. After that we will realize we have to include the dumper and pass 
> the array ref or hash ref.
> 
> It would be good we can dump the data structure with a single command.
> 
> *Describe the solution you'd like*
> 
> I like the pprint keyword in Python but there also you have to import it 
> before using it.
> For Perl if there is a keyword something like that can pretty print my 
> data structure without passing reference it would be good especially 
> during debug time. E.g.
> 
> |pprint @array; [ 'abc', 'ijk', 'xyz' ] pprint %hash or pprint 
> $output_fh %hash; { 'key1' => 'value1', 'key2' => 'value2' } pprint 
> $scalar; abc |
> 
> I have just use the 'pprint' keyword as an example. It can be 
> anything(e.g. 'printd' ?).
> 

Have you tried the Data::Dump module from CPAN?

**Comment by choroba** [comment]
[XXX](https://metacpan.org/pod/XXX) is another interesting module that makes it possible to print the structure even from inside a complex expression.

**Comment by Tux** [comment]
Might I promote my own [Data::Peek](https://github.com/Tux/Data-Peek/blob/master/doc/Peek.md)?
What you describe sounds like [P](https://metacpan.org/pod/P) which I would personally strongly advice **NOT** to use.

**Comment by Grinnz** [comment]
If I am understanding correctly this request is for a new builtin which will pass its arguments to Dumper and then to print, without having to load Data::Dumper, and not a request for any features which other CPAN modules may feature. This builtin could be added as a feature which could subsequently be included in a `use vX` feature bundle for ease of access.

With that in mind my thoughts are:

1. I don't think this is particularly useful enough to justify its own feature.
2. it needs a much better name than pprint.

**Comment by rai-gaurav** [comment]
Thanks all for the suggestions. 
Grinnz got my point.
I agree its trivial. That why it take some time for me to raise this.
I only felt the need after using the python print where we can see the internals of array and dict.

```
a = ["abc", "ijk", 1] 
print(a)          # ['abc', 'ijk', 1]

h = {'abc': 5, 'ijk': 7, 2: 10}
print(h)         # {'abc': 5, 'ijk': 7, 2: 10}
```
A similar thing in Perl will return the address (Though we can get the space separated array elements by double quoting it).
When doing debugging where I just want to see the internals quickly, I have to include some dump module before (which I have to remove later on also). In one liner also you have to include it for normal printing an array or hash.
The python one is good but when your data structure is big, it also become difficult to read and search for specific element, hence I have suggested for pretty print.

Its not something we can't live without. In fact still we are living without it :). 
I hope I am not the only one who thinks like that. If that's the case, then we can definitely close this topic and forgot about it. :P 

Regarding name, pprint was just an example. Few name I can suggest more -
1. "tell"  as tell me about this array
2. "reveal" as reveal this hash
3. "show" as show this hash.

**Comment by choroba** [comment]
[tell](http://p3rl.org/tell) is already taken.

**Comment by xiaoyafeng** [comment]
If you just wanna debug easily, you may try Devel::ptkdb which default loads Data::Dumper and show variable you choose.

**Comment by yuki-kimoto** [comment]
I also have a desire to dump easily, especially containing UTF-8 encoding.

I and my friend created a D module.

https://metacpan.org/pod/D

```
use utf8;

# Reference data that contains decoded strings
my $data = [{name => 'あ'}, {name => 'い'}];
 
# Examples of useful oneliner.
use D;du $data;
```

I don't think it's essential for the core,

on the other hand, I would be happy if this could be done by core module.

**Comment by Astara** [comment]
As @Tux suggests, 'P' handles this inline, automatically, but doesn't do the pretty-printing of Data::Dumper.

It was designed to be a more user-friendly print verb than any of the the existing options at various times.

As an example, using @yuki-kimoto's "my $data = [{name => 'あ'}, {name => 'い'}];":

>  perl -MP -e" P '%s', [{name => 'あ'}, {name => 'い'}]; "
[{name=>"あ"}, {name=>"い"}]

**Comment by rmallah** [comment]
how about 

```
use Data::Printer;

my %x = ( hello => 'world');
p %x ;
np %x;

```
https://metacpan.org/pod/Data::Printer

**Comment by Astara** [comment]
@rmallah Very pretty, though it doesn't work with the same type of input as the original user used:
> perl -MData::Printer -e " p [{name=> 'あ'}, {name => 'い'}]; "
Type of arg 1 to Data::Printer::p must be one of [@$%&] (not anonymous list ([])) at -e line 1, near "];"

though that shouldn't be hard to fix, no?  In 'p' just examine the reference to see its type, then use "[]" or "{}" around the original (or whatever works for a reference.  If it is for a class, put that before the interior class like: 
> perl -MP -e " P 'output=%s', bless [{name=> 'あ'}, {name => 'い'}],'yoClass'; "
output=yoClass[{name=>'あ'}, {name=>'い'}]

Wow...I would have thought '$' in the prototype would have handled [] or {}.  That's unfortunate.

**Comment by rmallah** [comment]
yes its literally checking the sigil :-) ,but in real debug scenarios  we deal mostly with variables
i use it frequently and I never felt the need to dump anonymous. 

np is quite handy with sprintf
